packages/runtime/src/plugins/generate-preload-assets.ts

```diff
@@ -0,0 +1,337 @@
+import {
+  GlobalModuleInfo,
+  ModuleInfo,
+  ProviderModuleInfo,
+  isManifestProvider,
+} from '@module-federation/sdk';
+import {
+  EntryAssets,
+  FederationRuntimePlugin,
+  PreloadAssets,
+  PreloadConfig,
+  PreloadOptions,
+  RemoteInfoOptionalVersion,
+} from '../type';
+import { assignRemoteInfo } from './snapshot';
+import { getInfoWithoutType, getPreloaded, setPreloaded } from '../global';
+import { getResourceUrl } from '../utils/manifest';
+import { FederationHost } from '../core';
+import { defaultPreloadArgs, normalizePreloadExposes } from '../utils/preload';
+import { getGlobalShare } from '../utils/share';
+import { getFMId, isPureRemoteEntry, isRemoteInfoWithEntry } from '../utils';
+
+declare global {
+  // eslint-disable-next-line no-var
+  var __INIT_VMOK_DEPLOY_GLOBAL_DATA__: boolean | undefined;
+}
+
+// name
+// name:version
+function splitId(id: string): {
+  name: string;
+  version: string | undefined;
+} {
+  const splitInfo = id.split(':');
+  if (splitInfo.length === 1) {
+    return {
+      name: splitInfo[0],
+      version: undefined,
+    };
+  } else if (splitInfo.length === 2) {
+    return {
+      name: splitInfo[0],
+      version: splitInfo[1],
+    };
+  } else {
+    return {
+      name: splitInfo[1],
+      version: splitInfo[2],
+    };
+  }
+}
+
+// Traverse all nodes through moduleInfo and traverse the entire snapshot
+function traverseModuleInfo(
```

Comment:
can module traversel be used outside of client runtime? 
Like could i traverse this on a edge or server independently to calculate module locks? 

Im guessing SnapshotHandler is designed for this use case

Reply:
yes , it could be . Now this function is for calculating all remote assets . 

if users set entry is 'remoteEntry.js' , the function will not work , because lack module infos . 

But if users set entry is 'xxx-manifest.json', snapshothandler will use this function get all we need assets .